spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/FPN.scala

```diff
@@ -0,0 +1,123 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class FPN[T : ClassTag](
+  val in_channels_list: Array[Int],
+  val out_channels: Int
+)
+  (implicit ev: TensorNumeric[T])
+  extends AbstractModule[Table, Table, T]{
+
+  private val num_feature_maps = in_channels_list.length
+  private val inner_blocks_modules = new Array[SpatialConvolution[T]](num_feature_maps)
+  private val layer_blocks_modules = new Array[SpatialConvolution[T]](num_feature_maps)
+
+  private val graph: Graph[T] = buildGraph()
+
+  def buildGraph(): Graph[T] = {
+    for (i <- 0 to num_feature_maps - 1) {
+      if (in_channels_list(i) != 0) {
+        val inner_block_module =
+          SpatialConvolution[T](in_channels_list(i), out_channels, 1, 1, 1, 1)
+        val layer_block_module =
+          SpatialConvolution[T](out_channels, out_channels, 3, 3, 1, 1, 1, 1)
+        inner_blocks_modules(i) = inner_block_module
+        layer_blocks_modules(i) = layer_block_module
+      }
+    }
+
+    val input = Input[T]()
+    val inner_conv = ParallelTable[T]()
+    for (i <- 0 to num_feature_maps - 1) {
+      inner_conv.add(inner_blocks_modules(i))
+    }
+    val inner_block = inner_conv.inputs(input)
+
+    var count = 0
+    var results = new Array[ModuleNode[T]](num_feature_maps)
+    var last_inner = SelectTable[T](num_feature_maps).inputs(inner_block)
+    results(count) = layer_blocks_modules(num_feature_maps - 1).inputs(last_inner)
+
+    for(i <- num_feature_maps - 1 to 1 by -1) {
+      val layer_block = layer_blocks_modules(i - 1)
+      if (layer_block != null) {
+        val inner_topdown = UpSampling2D[T](Array(2, 2)).inputs(last_inner)
+        val inner_lateral = SelectTable[T](i).inputs(inner_block)
+        last_inner = CAddTable[T]().inputs(inner_lateral, inner_topdown)
+        count += 1
+        results(count) = layer_block.inputs(last_inner)
+      }
+    }
+
+    Graph(Array(input), results)
+  }
+
+  override def updateOutput(input: Table): Table = {
+    output = graph.forward(input).toTable
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Table): Table = {
+    gradInput = null
```

Comment:
it should support backward

Reply:
Backward reports an error when Input and ParallelTable are connected in Graph. Changed some code in Graph. Now FPN backward works but not sure if the change has impacts on other modules.